### PR TITLE
EF Core Code Generation MVP

### PR DESCRIPTION
**Pull request by benmccallum**
Adds EF Core code generation and an analyzer so that we can go SDL-first --> EF model --> DB.

Functionality:
 * [x] Analyzer 
 * [x] Code generation project
 * [x] EF entities generation (from object types that will be backed by a table)
 * [x] EF DbContext generation (with DbSet members for object types that will be backed by a table)
 * [x] @schemaConventions directive (with pluralized table names option) (addresses cloud#116)
 * [x] @table directive (with name property to specify custom table name mapping) (addresses cloud#117)
 * [x] PK determination (@key, falling back to Id-looking field, falling back to auto-created field using schema convention) (addresses cloud#112)
 * [x] A bunch of other stubbed directives that don't do anything yet
 * [x] GQL `Query` type generation
 * [ ] Field resolver per table entity (with IQueryable, x attr, y attr, z attr)
 * [x] Startup DI code generation

Concepts:
* As a schema object type is processed, a context is built up that allows us to capture things like PK, table name, etc and build up our knowledge of the SDL as we need to generate code. This will help us in the future when we start doing more complicated directives for things like relations.

Testing:
 * [x] Snapshooter test coverage of the raw code generated
 * [ ] Integration testing (as per Neo4J, that also compiles the source) (likely another PR)

### CODE REVIEW

**Comment by benmccallum** [comment]
@michaelstaib , this is ready for review really. Tomorrow I'm gonna:
* add a couple of tests for the PK stuff and 
* add the code generation for the DI part, which I'll try copy the solution from the Neo4J one for

But you should start reviewing it in any case if you have the time.

File: src/HotChocolate/Analyzers/src/Analyzers/EntityFrameworkSourceGenerator.TypeInitialization.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+using System;
+using System.Reflection;
+using static System.IO.Path;
+
+namespace HotChocolate.Analyzers
+{
+    public partial class EntityFrameworkSourceGenerator
+    {
+        private const string _dll = ".dll";
```
Comment (michaelstaib) [inline src/HotChocolate/Analyzers/src/Analyzers/EntityFrameworkSourceGenerator.TypeInitialization.cs:9]: I think we do not need this once I have merged my generator improvements.

File: src/HotChocolate/CodeGeneration/src/CodeGeneration.EntityFramework/EntityFrameworkSchemaHelper.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            schemaBuilder.AddDirectiveType<JsonDirectiveType>();
+            schemaBuilder.AddDirectiveType<KeyDirectiveType>();
+            schemaBuilder.AddDirectiveType<ForeignKeyDirectiveType>();
+            schemaBuilder.AddDirectiveType<IndexDirectiveType>();
+            schemaBuilder.AddDirectiveType<OneToOneDirectiveType>();
+            schemaBuilder.AddDirectiveType<OneToManyDirectiveType>();
+            schemaBuilder.AddDirectiveType<ManyToManyDirectiveType>();
+
+            // EntityFramework
+
```
Comment (michaelstaib) [inline src/HotChocolate/CodeGeneration/src/CodeGeneration.EntityFramework/EntityFrameworkSchemaHelper.cs:37]:
```suggestion
            // EntityFramework
```

Comment (michaelstaib) [reply src/HotChocolate/CodeGeneration/src/CodeGeneration.EntityFramework/EntityFrameworkSchemaHelper.cs:37]: line

File: src/HotChocolate/CodeGeneration/src/CodeGeneration.EntityFramework/Types/ForeignKeyDirective.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+using HotChocolate.Types;
+
+namespace HotChocolate.CodeGeneration.EntityFramework.Types
+{
+    public class ForeignKeyDirective
+    {
+        public string To { get; set; } = default!; // TODO: Potentially this supports nullable and can be derived by the field it annotates
```
Comment (michaelstaib) [inline src/HotChocolate/CodeGeneration/src/CodeGeneration.EntityFramework/Types/ForeignKeyDirective.cs:7]: Once I reworked directives we could also have a constructor that passes this in.

File: src/HotChocolate/CodeGeneration/src/CodeGeneration.EntityFramework/Types/ManyToManyDirective.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    public class ManyToManyDirectiveType : DirectiveType<ManyToManyDirective>
+    {
+        protected override void Configure(IDirectiveTypeDescriptor<ManyToManyDirective> descriptor)
+        {
+            descriptor
+                .Name("manyToMany")
+                .Location(DirectiveLocation.FieldDefinition);
+
+            descriptor
+               .Argument(t => t.JoinTable)
```
Comment (michaelstaib) [inline src/HotChocolate/CodeGeneration/src/CodeGeneration.EntityFramework/Types/ManyToManyDirective.cs:31]: the indention seems to be off

File: src/HotChocolate/CodeGeneration/src/CodeGeneration.EntityFramework/Types/OneToManyDirective.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+using HotChocolate.Types;
+
+namespace HotChocolate.CodeGeneration.EntityFramework.Types
+{
+    public class OneToManyDirective
```
Comment (michaelstaib) [inline src/HotChocolate/CodeGeneration/src/CodeGeneration.EntityFramework/Types/OneToManyDirective.cs:5]: It is nice that you collocate the type wit the model.

**Review by michaelstaib — APPROVED** [review]

File: src/HotChocolate/Core/src/Types/Types/Contracts/IDirective.cs [hunk 6]
```diff
@@ -17,6 +18,8 @@ public interface IDirective
 
         T ToObject<T>();
 
+        T ToObject<T>(Type type);
```
Comment (benmccallum) [inline src/HotChocolate/Core/src/Types/Types/Contracts/IDirective.cs:21]: @michaelstaib , would this be a problem?

I have a scenario where I'd like to be able to loop the directives of a field, finding those that implement an interface, but then actually get the directive class, but `.ToObject<T>` doesn't cut it as T in my case would be an interface and the Activator code fails. Have a look at how it's used in `SchemaExtensions::GetDirectivesWhereRuntimeTypeImplements`

Comment (benmccallum) [reply src/HotChocolate/Core/src/Types/Types/Contracts/IDirective.cs:21]: Usage:
`directive.ToObject<IXyzDirective>(directive.Type.RuntimeType)`

**Pull request closed without merging**